File: forge/src/main/java/dev/kineticcat/hexportation/forge/transfer/ItemVariantImpl.java
package dev.kineticcat.hexportation.forge.transfer;

import dev.kineticcat.hexportation.api.transfer.ItemVariant;
import net.minecraft.world.item.Item;
import net.minecraft.world.item.ItemStack;

/**
 * Forge implementation using singleton injection pattern.
 */
public class ItemVariantImpl extends ItemVariant.Implementation {
    
    public static final ItemVariantImpl INSTANCE = new ItemVariantImpl();
    
    @Override
    public ItemVariant of(Item item) {
        return new ItemStackWrapper(new ItemStack(item));
    }
    
    @Override
    public ItemVariant blank() {
        return new ItemStackWrapper(ItemStack.EMPTY);
    }
    
    /**
     * Wrapper class that extends our ItemVariant abstract class using Forge's ItemStack.
     */
    public static class ItemStackWrapper extends ItemVariant {
        private final ItemStack itemStack;
        
        public ItemStackWrapper(ItemStack itemStack) {
            this.itemStack = itemStack;
        }
        
        @Override
        public Item getItem() {
            return itemStack.getItem();
        }
        
        @Override
        public ItemStack toStack() {
            return itemStack.copy();
        }
        
        @Override
        public ItemStack toStack(int count) {
            ItemStack copy = itemStack.copy();
            copy.setCount(count);
            return copy;
        }
        
        public ItemStack getItemStack() {
            return itemStack;
        }
        
        @Override
        public boolean equals(Object obj) {
            if (this == obj) return true;
            if (!(obj instanceof ItemVariant)) return false;
            
            // Compare items - for Forge we only care about the Item type, not NBT for now
            if (obj instanceof ItemStackWrapper other) {
                return this.itemStack.getItem() == other.itemStack.getItem();
            }
            
            // Generic ItemVariant comparison
            ItemVariant other = (ItemVariant) obj;
            return this.getItem() == other.getItem();
        }
        
        @Override
        public int hashCode() {
            return itemStack.getItem().hashCode();
        }
    }
}

File: forge/src/main/java/dev/kineticcat/hexportation/forge/transfer/ItemVariantImpl.kt
package dev.kineticcat.hexportation.forge.transfer

import dev.kineticcat.hexportation.api.transfer.ItemVariant
import net.minecraft.world.item.Item
import net.minecraft.world.item.ItemStack

/**
 * Forge implementation using singleton injection pattern.
 */
class ItemVariantImpl : ItemVariant.Implementation() {
    
    companion object {
        @JvmField
        val INSTANCE = ItemVariantImpl()
    }
    
    override fun of(item: Item): ItemVariant = ItemStackWrapper(ItemStack(item))
    
    override fun blank(): ItemVariant = ItemStackWrapper(ItemStack.EMPTY)
    
    /**
     * Wrapper class that extends our ItemVariant abstract class using Forge's ItemStack.
     */
    class ItemStackWrapper(
        private val itemStack: ItemStack
    ) : ItemVariant() {
        
        override val item: Item get() = itemStack.item
        
        override fun toStack(): ItemStack = itemStack.copy()
        
        override fun toStack(count: Int): ItemStack = itemStack.copy().apply { setCount(count) }
        
        fun getItemStack() = itemStack
        
        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (other !is ItemVariant) return false
            
            // Compare items - for Forge we only care about the Item type, not NBT for now
            return when (other) {
                is ItemStackWrapper -> itemStack.item == other.itemStack.item
                else -> item == other.item
            }
        }
        
        override fun hashCode(): Int = itemStack.item.hashCode()
    }
}